Add theme compatibility for Course Page

alexsanford opened the pull request:
Partial solution to #2154

Adds initial architecture for rendering within unsupported themes. Also provides full implementation for the Course Page.

## Implementation notes

### Renderers

I've made the architectural decision to move the rendering of the various pages into Renderer classes. This is different from WooCommerce's solution, which generally renders shortcodes. My thought is that having a renderer class whose sole purpose is to render the content can be helpful for rendering pages as we are now, rendering shortcodes, and later rendering blocks used by Gutenberg. Note that the `sensei_course_page` shortcode's functionality has been simply moved into the Renderer.

If it seems like this architecture is going to add too much complexity, then I can go back to just using the shortcode, but note that we will still have to manually render (or create a shortcode for) all of the pages that do not yet have shortcodes.

### Theme support

Theme support for Sensei is declared by calling `add_theme_support( 'sensei' );`. If theme support is added, then the changes to rendering added by this PR are disregarded. They are also disregarded if the current theme is one that Sensei supports (see [`includes/theme-integrations/theme-integration-loader.php`](https://github.com/Automattic/sensei/blob/c181c59be42e83e382823d4a142360da1255a9d2/includes/theme-integrations/theme-integration-loader.php)).

### Reviewing

The most important changes were made in the first two commits. The rest were for fixing some CSS issues that cropped up on some themes, and ensuring the video showed up on the course page.

## Concerns

- I've modified the shortcode rendering to include the video embed on a single-course page. It appears that this was not being rendered in the shortcode before. I'm not sure whether this change will cause a problem for existing users. If so, we could add a parameter to the shortcode to turn off video rendering. I'm thinking it's probably not necessary, as it makes more sense for the video to be rendered.

- I've made some CSS tweaks based on some small issues I've found in a few themes I used for testing. More such tweaks may be necessary as we test on more themes.

- There is a problem with themes that add their own "next/previous" pagination links. Since the Sensei Course Page also provides these pagination links, this causes a doubling of pagination links at the bottom of the page. Specifically I saw this issue in the following themes:
  - Sydney
  - Mesmerize
  - Ashe

- There are no unit tests. We may be able to add some for the Renderers?

## Testing

To test, simply load a course page on several different themes. Make sure to try themes that support Sensei, and themes that do not. Also try with themes that Sensei provides integration for, and compare it to a page which renders the shortcode: `[sensei_course_page id=<course-id>]`

The most important, however, are the themes that do not support Sensei at all. We want to make sure that this PR prevents the layout from breaking, and makes the course page look decent (it is expected that there may still be some small CSS issues).

donnapep commented:
> There is a problem with themes that add their own "next/previous" pagination links.

Is there anything we can do here, or would we need to perhaps provide users with some sort of snippet that they could add to `functions.php` to fix it?

> I've made some CSS tweaks based on some small issues I've found in a few themes I used for testing. More such tweaks may be necessary as we test on more themes.

If someone is using the same CSS selectors that we were, will they need to make the same CSS changes?

> There are no unit tests. We may be able to add some for the Renderers?

That would be ideal. I almost think we should have a policy of not approving a PR unless it has sufficient test coverage.

alexsanford commented:
> Is there anything we can do here, or would we need to perhaps provide users with some sort of snippet that they could add to functions.php to fix it?

I'd have to take a closer look, but my gut feeling is that we wouldn't be able to handle this automatically. We should, however, be able to pretty easily set up a filter so that the user can turn off our pagination in `functions.php`.

> If someone is using the same CSS selectors that we were, will they need to make the same CSS changes?

At this point, probably not. But in general, yeah we may need them to make CSS changes.

In this PR, I changed three things in the CSS:

- Removed references to the `#main` element, which may not be there anymore.
- Changed the `.single-course` selector to `.course`, which matches the element whether it comes from the template or the shortcode.
- Added a few margins to give some spacing between elements on the Course page.

I'm thinking that if the user has custom CSS that is working, probably it will continue to work after this change. Especially since if their site is working, they should probably just disable unsupported rendering anyway (which we will need to provide them instructions for).

> That would be ideal. I almost think we should have a policy of not approving a PR unless it has sufficient test coverage.

Sounds like a great idea. I'll make sure to get some test coverage set up for when we dive back into this for next release.

alexsanford commented:
I rebased and force-pushed this. The additions:

- Added a filter in https://github.com/Automattic/sensei/pull/2155/commits/8e41d94c25a6ac8db945041447e9d3f3ad73c670 to disable showing pagination on the course page. This is useful if the theme supplies its own pagination.

- Added unit tests for the new functionality introduced by this PR.

@jom @donnapep This is ready for another look when you have a chance! 🙂

alexsanford commented:
Thanks @jom for the review and testing!

I made the suggested changes 🙂

alexsanford commented:
@jom FYI I did a small refactor of the Renderer in https://github.com/Automattic/sensei/pull/2155/commits/02b8001bdabfb37c169f586b692b3c2a9f3a18ed. It's not huge but it does touch a bunch of files. Could you take a look at that refactor and let me know what you think? Maybe it's unnecessary, but I think it is more semantic and cleans things up a bit (no more need for the `Sensei_Renderer_Missing_Fields_Exception` for example).

alexsanford commented:
Rebased ^

Will merge when CI passes.

alexsanford commented:
Because the CI failure is an issue with the CI setup (see https://github.com/Automattic/sensei/pull/2207), particularly in the JS testing environment, this is safe to merge despite the failure.
